Error in concrete intersection of HPolygon

```julia
julia> h = Hyperrectangle([0.9806984373998077, 0.8502036865893472], [0.002214407802010432, 0.0007748971207262034])
Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}([0.9806984373998077, 0.8502036865893472], [0.002214407802010432, 0.0007748971207262034])

julia> B = Ball2([1.0, 1.0], 0.15)
Ball2{Float64,Array{Float64,1}}([1.0, 1.0], 0.15)

julia> p = overapproximate(B, 1e-4);

julia> isempty(LazySets.intersection(HPolygon(constraints_list(h)), p))
true

julia> isempty(intersection(h, p))
false

julia> LazySets.isdisjoint(h, p)
false

julia> any(v -> v ∈ p, vertices_list(h))
true
```

Related to this function

```julia
A = rand(Zonotope)
X = rand(Zonotope)
intersection(overapproximate(X + A, 1e-3), overapproximate(X + B, 1e-3))

AssertionError: the constraints must have the same normal direction

Stacktrace:
 [1] macro expansion at /home/mforets/.julia/dev/LazySets/src/Assertions/Assertions.jl:23 [inlined]
 [2] is_tighter_same_dir_2D(::HalfSpace{Float64,Array{Float64,1}}, ::HalfSpace{Float64,Array{Float64,1}}; strict::Bool) at /home/mforets/.julia/dev/LazySets/src/Sets/HalfSpace.jl:460
 [3] is_tighter_same_dir_2D at /home/mforets/.julia/dev/LazySets/src/Sets/HalfSpace.jl:459 [inlined]
 [4] intersection(::HPolygon{Float64,Array{Float64,1}}, ::HPolygon{Float64,Array{Float64,1}}, ::Bool) at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/intersection.jl:515
 [5] intersection(::HPolygon{Float64,Array{Float64,1}}, ::HPolygon{Float64,Array{Float64,1}}) at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/intersection.jl:469
 [6] top-level scope at In[134]:3

```

--- a/src/ConcreteOperations/intersection.jl
+++ b/src/ConcreteOperations/intersection.jl
@@ -586,7 +586,9 @@ function intersection(P1::AbstractHPolygon{N},
         deleteat!(c, length(c)-duplicates+1:length(c))
     end
 
-    P = HPolygon(c, sort_constraints=false)
+    # TODO see #2187: the above code does *not* sort the constraints correctly.
+    # after fixing the above code, we should pass sort_constraints=false again
+    P = HPolygon(c, sort_constraints=true)
     if prune
         remove_redundant_constraints!(P)
         if isempty(P)

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -518,6 +518,12 @@ for N in [Float64, Float32]
     addconstraint!(p2, p2.constraints[2])
     @test length(p2.constraints) == 6
 
+    # correct sorting of constraints in intersection (#2187)
+    h = Hyperrectangle(N[0.98069, 0.85020],
+                       N[0.00221, 0.00077])
+    p = overapproximate(Ball2(N[1, 1], N(15//100)), 1e-4)
+    @test !isempty(intersection(HPolygon(constraints_list(h)), p))
+
     # test that concrete minkowski sum of a singleton and a polygon works
     x = VPolygon([N[1, 1]])
     y = VPolygon([N[1.1, 2.2], N[0.9, 2.2], N[0.9, 2.0], N[1.1, 2.0]])

